File: logiflow/delivery-graphql-service/src/services/billing.client.ts
import { billingClient } from '../utils/axiosClients';

/**
 * Respuesta de factura del billing-service
 */
export interface FacturaResponse {
  id: string;
  pedidoId: string;
  tipoEntrega: string;
  montoTotal: number;
  estado: string;
  fechaCreacion: string;
  distanciaKm: number;
}

/**
 * Cliente para el billing-service
 * Endpoints:
 * - GET /billing/facturas/{id} -> Factura por ID
 * - GET /billing/facturas/pedido/{pedidoId} -> Factura por pedidoId
 */
export class BillingServiceClient {

  /**
   * Obtener factura por ID
   */
  async obtenerFacturaPorId(id: string): Promise<FacturaResponse | null> {
    try {
      console.log(`[BillingClient] GET /billing/facturas/${id}`);
      const response = await billingClient.get<FacturaResponse>(`/billing/facturas/${id}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        console.log(`[BillingClient] Factura ${id} no encontrada`);
        return null;
      }
      console.error(`[BillingClient] Error obteniendo factura ${id}:`, error.message);
      throw error;
    }
  }

  /**
   * Obtener factura por pedidoId
   */
  async obtenerFacturaPorPedidoId(pedidoId: string): Promise<FacturaResponse | null> {
    try {
      console.log(`[BillingClient] GET /billing/facturas/pedido/${pedidoId}`);
      const response = await billingClient.get<FacturaResponse>(`/billing/facturas/pedido/${pedidoId}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        console.log(`[BillingClient] Factura para pedido ${pedidoId} no encontrada`);
        return null;
      }
      console.error(`[BillingClient] Error obteniendo factura de pedido ${pedidoId}:`, error.message);
      throw error;
    }
  }
}

File: logiflow/pedido-service/src/main/java/com/logiflow/pedidoservice/client/BillingClient.java
package com.logiflow.pedidoservice.client;

import com.logiflow.pedidoservice.dto.FacturaRequest;
import com.logiflow.pedidoservice.dto.FacturaResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

/**
 * Cliente REST para comunicación con Billing Service
 * Usa RestTemplate (Spring tradicional)
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class BillingClient {

    private final RestTemplate restTemplate;

    @Value("${services.billing.url:http://localhost:8082}")
    private String billingServiceUrl;

    /**
     * Crea una factura en el Billing Service
     *
     * @param request datos del pedido para calcular la tarifa
     * @param token
     * @return respuesta con la factura creada
     */


        public FacturaResponse crearFactura(FacturaRequest request, String token) {
            try {
                log.info("Llamando a Billing Service para crear factura - pedidoId: {}", request.getPedidoId());

                String url = billingServiceUrl + "/api/facturas";

                // 🔹 CONFIGURACIÓN DE HEADERS
                HttpHeaders headers = new HttpHeaders();

                // Verificamos si el token ya trae la palabra "Bearer "
                String tokenHeader = token.startsWith("Bearer ") ? token : "Bearer " + token;
                headers.set("Authorization", tokenHeader);

                // 🔹 Envolvemos request + headers
                HttpEntity<FacturaRequest> entity = new HttpEntity<>(request, headers);

                // 🔹 Hacemos el POST (Cambiamos postForObject por exchange para mayor control si fuera necesario)
                FacturaResponse response = restTemplate.postForObject(
                        url,
                        entity,
                        FacturaResponse.class
                );

                log.info("💰 [BILLING-CLIENT] Factura creada exitosamente | FacturaID: {} | PedidoID: {} | Monto: ${}", 
                    response.getId(), request.getPedidoId(), response.getMontoTotal());
                log.info("🔗 [FACTURA-RESPONSE] Cliente recibió respuesta | FacturaID: {} para PedidoID: {}",
                    response.getId(), request.getPedidoId());
                return response;

            } catch (Exception e) {
                log.error("Error al comunicarse con Billing Service: {}", e.getMessage());
                throw new RuntimeException("No se pudo crear la factura: " + e.getMessage(), e);
            }
        }

    /**
     * Obtiene una factura por ID del pedido
     *
     * @param pedidoId ID del pedido (UUID como String)
     * @return respuesta con la factura
     */
    public FacturaResponse obtenerFacturaPorPedidoId(String pedidoId) {
        try {
            log.info("Consultando factura por pedidoId: {}", pedidoId);

            String url = billingServiceUrl + "/api/facturas/pedido/" + pedidoId;

            FacturaResponse response = restTemplate.getForObject(url, FacturaResponse.class);

            log.info("Factura encontrada - facturaId: {}", response.getId());

            return response;

        } catch (RestClientException e) {
            log.error("Error al obtener factura por pedidoId: {}", e.getMessage());
            return null; // Devuelve null si no se encuentra
        }
    }
}

